src/snowflake/connector/connection_diagnostic.py

```diff
@@ -0,0 +1,365 @@
+        if not self.report_destination.is_absolute():
+            path_failed = True
+            logger.warning(f"Path({self.report_destination} for connection test is not absolute and is not valid.")
+        elif not self.report_destination.exists():
+            path_failed = True
+            logger.warning(f"Path({self.report_destination} for connection test does not exist.")
+
+        if path_failed:
+            if IS_WINDOWS:
+                new_report_destination = Path(tempfile.gettempdir())
+            else:
+                new_report_destination = Path("/tmp/")
+            logger.warning(f"Since the provided path({self.report_destination}) was invalid, using"
+                           f" {new_report_destination}")
+            self.report_destination = new_report_destination
+
+        self.report_file = Path(f"{str(self.report_destination)}/{self.report_file}")
+        logger.info(f"Reporting to file {self.report_file}")
+
+        self.whitelist_json = Path(str(kwargs.get('connection_diag_whitelist_path')))
+        if not self.whitelist_json.is_absolute():
+            logger.warning(f"Path '{self.whitelist_json}' for connection test whitelist is not absolute and is not "
+                           f"valid.")
+            logger.warning(f"Will connect to Snowflake for whitelist json instead.  If you did not provide a valid "
+                           f"password, please make sure to update and run again.")
+            self.whitelist_json = None
+        elif not self.whitelist_json.exists():
+            logger.warning(f"File '{self.report_destination}' for connection test whitelist does not exist.")
+            logger.warning(f"Will connect to Snowflake for whitelist json instead.  If you did not provide a valid "
+                           f"password, please make sure to update and run again.")
+            self.whitelist_json = None
+
+        self.initial_host = kwargs.get("host")
+        self.host = self.initial_host
+        self.account = kwargs.get("account")
+        self.proxy_host = kwargs.get("proxy_host")
+        self.proxy_port = kwargs.get("proxy_port")
+        self.proxy_user = kwargs.get("proxy_user")
+        self.proxy_password = kwargs.get("proxy_password")
+        self.test_results = {'INITIAL': [], 'PROXY': [], 'SNOWFLAKE_URL': [], 'STAGE': [], 'OCSP_RESPONDER': [],
+                             'OUT_OF_BAND_TELEMETRY': []}
+        self.ocsp_urls = []
+        if self.__is_privatelink():
+            self.ocsp_urls.append(f"ocsp.{self.host}")
+            self.whitelist_sql = "select system$whitelist_privatelink();"
+        else:
+            self.ocsp_urls.append(f"ocsp.snowflakecomputing.com")
+            self.whitelist_sql = "select system$whitelist();"
+        self.whitelist_retrieval_success = False
+        self.crl_urls = []
+        self.cursor = None
+
+    def set_cursor(self, cursor):
+        self.cursor = cursor
```

Comment:
Let's not use Java programming.

Add this to the initalizer:
```python
self.cursor: SnowflakeCursor | None = None
```
Then just set the parameter instead of calling this function